components/web_discovery/browser/content_scraper.h

```diff
@@ -0,0 +1,116 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMPONENTS_WEB_DISCOVERY_BROWSER_CONTENT_SCRAPER_H_
+#define BRAVE_COMPONENTS_WEB_DISCOVERY_BROWSER_CONTENT_SCRAPER_H_
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "base/containers/flat_map.h"
+#include "base/functional/callback.h"
+#include "base/values.h"
+#include "brave/components/web_discovery/browser/document_extractor/lib.rs.h"
+#include "brave/components/web_discovery/browser/patterns.h"
+#include "brave/components/web_discovery/browser/server_config_loader.h"
+#include "brave/components/web_discovery/common/web_discovery.mojom.h"
+#include "url/gurl.h"
+
+namespace web_discovery {
+
+struct PageScrapeResult {
+  PageScrapeResult(GURL url, std::string id);
+  ~PageScrapeResult();
+
+  PageScrapeResult(const PageScrapeResult&) = delete;
+  PageScrapeResult& operator=(const PageScrapeResult&) = delete;
+
+  base::Value SerializeToValue();
+  static std::unique_ptr<PageScrapeResult> FromValue(const base::Value& dict);
+
+  GURL url;
+  // A map of DOM selectors to list of scraped values embedded in a Dict.
+  // Each dict contains arbitrary keys (defined in the patterns) to scraped
+  // values.
+  base::flat_map<std::string, std::vector<base::Value::Dict>> fields;
```

Comment:
This would have been simpler, and easier to test if it had been just:

```cxx
base::flat_map<std::string, base::Value::List> fields;
```

There is a semantical difference in there, but in practice the code is easier to read.